crates/circuit/src/dag_circuit.rs

```diff
@@ -1634,13 +1670,35 @@ impl DAGCircuit {
             })
             .transpose()?;
 
+        // TODO: since operations own their blocks in Python land, the best we can do without extra
+        //       tracking is to create a new block in 'self' for every block in 'other'. This will
+        //       lead to duplication if a Python user composes with a DAG that already uses blocks
+        //       registered in 'self'.
```

Comment:
Anything in Rust that introduces multiple references to the same `Block` from control-flow operations is already going to break routing and other parts of the transpiler anyway right now.